rfcs/2021-03-26-6517-end-to-end-acknowledgement.md

```diff
@@ -0,0 +1,279 @@
+
+### Event metadata
+
+Given the channel above, the following metadata will be added to events:
+
+1.  A reference (`Arc`) to a originating source, which is a dual-purpose structure containing:
+
+    1.  The finalization channel, to be used when actually delivering the status.
+
+    2.  The unique source identifier, to be used when serializing the metadata.
+
+2.  A event identifier, contained in a new enum type.
+
+
+### Data structures
+
+Since a given event may actually be a composition of multiple source
+events, this must be a set of the above pairs of data.
+
+```rust
+enum EventId {
+    Number(u64),
+    SenderNumber(u64, u64),
+    // additional variants as required
+}
+
+enum EventStatus {
+    Delivered,
+    Dropped,
+    Failed,
+}
+
+struct EventFinalization {
+    id: EventId,
+    status: EventStatus,
+}
+
+struct OriginatingSource {
+    receiver: tokio::sync::mpsc::Receiver<EventFinalization>,
+    identifier: String,
+}
+
+struct EventSource {
+    source: Arc<OriginatingSource>,
+    id: EventId,
+}
+
+struct EventMetadata {
+    // existing fields
+    sources: Box<\[EventSource\]>,
+}
+```
+
+## Doc-level Proposal
+
+The following additional source configuration will be added:
+
+```toml
+# Global enable option
+acknowledgements = true
+
+[sources.my_source_id]
+  # Per-source enable option
+  acknowledgements = true
+```
+
+We would also need to document when acknowledgement happens for each source.
+
+## Rationale
+
+The above structure provides for several considerations:
+
+1.  This the minimum amount of data that can be added to the metadata to
+    fully support this feature.
+
+2.  Each source element is fixed size and requires no additional
+    allocations.
+
+3.  If a source does not need or is not configured to require
+    finalization, it will not contribute to the list of sources and so
+    has no additional event overhead, and no additional allocations when
+    the event is created.
+
+4.  Sending the finalization status to the source does not require any
+    lookups or topology traversal.
+
+5.  No additional work is required to handle dropped sources due to
+    topology reconfiguration, other than the expected checking for a
+    closed channel when sending.
+
+## Drawbacks
+
+1.  This adds a substantial base size overhead to each event (minimum of
+    two words), even for configurations that do not support or require
+    end-to-end acknowledgement.
+
+## Alternatives
+
+1.  The set of sources could be stored in a more customary
+    `Vec<EventSource>`. This provides for merging multiple sources with
+    a minimum of code. However, the data added to the event structure
+    then grows by an additional word.
+
+2.  The originating source could be stored as simply the unique
+    identifier string. This requires that all reporting of delivery
+    status proceed through a dictionary lookup instead of simply sending
+    it through a channel, increasing the run-time overhead.
```

Comment:
It would be wise to wrap the acknowledgement handling in its own struct like `Acker`, and we would definitely want to amortize costs as much as possible. The problem is that each output batch could be composed of events from a mix of sources, possibly interleaved or otherwise mixed, making simple start+length accounting very messy at best. (and that doesn't even take into account things like the metrics batch buffer)